File: leetcode/String/leetcode-0003/3.go
package string

func lengthOfLongestSubstring(s string) int {
	if len(s) == 0 {
		return 0
	}
	res := 0
	m := make(map[byte]int)
	i, j := 0, 0
	for i < len(s) && j < len(s) {
		for val, ok := m[s[j]]; ok && val >= i; i++ {
			m[s[i]] = -1
		}
		m[s[j]] = j
		j++
		res = max(res, j-i)
	}
	return res
}

func max(x, y int) int {
	if x > y {
		return x
	}
	return y
}

File: leetcode/String/leetcode-0003/3.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        if (s == null || s.length() == 0) {
            return 0;
        }
        int res = 0;
        HashSet<Character> set = new HashSet<>();
        int i = 0, j = 0;
        while (i < s.length() && j < s.length()) {
            while (set.contains(s.charAt(j))) {
                set.remove(s.charAt(i++));
            }
            set.add(s.charAt(j++));
            res = Math.max(res, j - i);
        }
        return res;
    }
}